components/brave_wallet/browser/zcash/zcash_orchard_sync_state.cc

```diff
@@ -0,0 +1,147 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_orchard_sync_state.h"
+
+#include <utility>
+
+#include "base/check_is_test.h"
+#include "base/containers/extend.h"
+#include "brave/components/brave_wallet/browser/zcash/orchard_shard_tree_delegate_impl.h"
+#include "brave/components/brave_wallet/common/zcash_utils.h"
+
+namespace brave_wallet {
+
+ZCashOrchardSyncState::ZCashOrchardSyncState(base::FilePath path_to_database) {
+  storage_ = base::MakeRefCounted<ZCashOrchardStorage>(path_to_database);
+}
+
+ZCashOrchardSyncState::~ZCashOrchardSyncState() {}
+
+// static
+void ZCashOrchardSyncState::OverrideShardTreeManagerForTesting(
+    const mojom::AccountIdPtr& account_id,
+    std::unique_ptr<OrchardShardTreeManager> manager) {
+  CHECK_IS_TEST();
+  shard_tree_managers_[account_id.Clone()] =
+      OrchardShardTreeManager::CreateForTesting(
+          std::make_unique<OrchardShardTreeDelegateImpl>(account_id.Clone(),
+                                                         storage_));
+}
+
+OrchardShardTreeManager* ZCashOrchardSyncState::GetOrCreateShardTreeManager(
+    const mojom::AccountIdPtr& account_id) {
+  if (shard_tree_managers_.find(account_id) == shard_tree_managers_.end()) {
+    shard_tree_managers_[account_id.Clone()] = OrchardShardTreeManager::Create(
+        std::make_unique<OrchardShardTreeDelegateImpl>(account_id.Clone(),
+                                                       storage_));
+  }
+  return shard_tree_managers_[account_id.Clone()].get();
+}
+
+base::expected<ZCashOrchardStorage::AccountMeta, ZCashOrchardStorage::Error>
+ZCashOrchardSyncState::RegisterAccount(mojom::AccountIdPtr account_id,
+                                       uint64_t account_birthday_block) {
+  return storage_->RegisterAccount(std::move(account_id),
```

Comment:
I don't quite understand the purpose of this class when most of the methods are just forwarding to another class. It doesn't seem like it has a well defined responsibility and doesn't have any comments explaining the class

Reply:
this is also missing unit tests

Reply:
It describes synchronization state. This includes discovered and spent notes plus commitment tree.
For some methods it uses database directly and for other methods it forwards calls to the OrchardShardTreeManager which then calls back to the database.

Reply:
We also need to place OrchardShardTreeManager and ZCashOrchardStorage into one place since they should be called on the same sequence.